terraform/context.go

```diff
@@ -119,24 +120,106 @@ func NewContext(opts *ContextOpts) (*Context, error) {
 		par = 10
 	}
 
-	// Setup the variables. We first take the variables given to us.
-	// We then merge in the variables set in the environment.
+	// Set up the variables in the following sequence:
+	//    0 - Take default values from the configuration
+	//    1 - Take values from TF_VAR_x environment variables
+	//    2 - Take values specified in -var flags, overriding values
+	//        set by environment variables if necessary.
 	variables := make(map[string]interface{})
-	for _, v := range os.Environ() {
-		if !strings.HasPrefix(v, VarEnvPrefix) {
-			continue
+
+	if opts.Module != nil {
+		for _, v := range opts.Module.Config().Variables {
+			if v.Default != nil {
+				if v.Type() == config.VariableTypeString {
+					// v.Default has already been parsed as HCL so there may be
+					// some stray ints in there
+					switch typedDefault := v.Default.(type) {
+					case string:
+						if typedDefault == "" {
+							continue
+						}
+						variables[v.Name] = typedDefault
+					case int:
+						variables[v.Name] = fmt.Sprintf("%d", typedDefault)
+					case bool:
+						variables[v.Name] = fmt.Sprintf("%t", typedDefault)
+					}
```

Comment:
Ah this is sadness. Totally understand why it is necessary. There might be some edge cases here (do big ints become int64? are floats valid?)